### PR TITLE
Neural network optimizers module

### PR DESCRIPTION
**Pull request by shretadas**
<html>
<body>
<h2 data-start="103" data-end="138">Neural Network Optimizers Module</h2>
<p data-start="140" data-end="444">This PR introduces a comprehensive neural network optimizers module that implements five widely used optimization algorithms for machine learning and deep learning.<br data-start="304" data-end="307">
The primary goal is to enhance the educational value of the repository by including well-documented, tested, and modular implementations.</p>
<p data-start="446" data-end="458">Fixes #13662</p>
<hr data-start="460" data-end="463">
<h3 data-start="465" data-end="481">What's Added</h3>
<ul data-start="483" data-end="1062">
<li data-start="483" data-end="538">
<p data-start="485" data-end="538">Add <strong data-start="489" data-end="496">SGD</strong> (Stochastic Gradient Descent) optimizer</p>
</li>
<li data-start="539" data-end="589">
<p data-start="541" data-end="589">Add <strong data-start="545" data-end="560">MomentumSGD</strong> with momentum acceleration</p>
</li>
<li data-start="590" data-end="647">
<p data-start="592" data-end="647">Add <strong data-start="596" data-end="603">NAG</strong> (Nesterov Accelerated Gradient) optimizer</p>
</li>
<li data-start="648" data-end="696">
<p data-start="650" data-end="696">Add <strong data-start="654" data-end="665">Adagrad</strong> with adaptive learning rates</p>
</li>
<li data-start="697" data-end="754">
<p data-start="699" data-end="754">Add <strong data-start="703" data-end="711">Adam</strong> optimizer combining momentum and RMSprop</p>
</li>
<li data-start="755" data-end="806">
<p data-start="757" data-end="806">Include 61 comprehensive doctests (all passing)</p>
</li>
<li data-start="807" data-end="866">
<p data-start="809" data-end="866">Add abstract <code data-start="822" data-end="837">BaseOptimizer</code> for a consistent interface</p>
</li>
<li data-start="867" data-end="914">
<p data-start="869" data-end="914">Include detailed mathematical documentation</p>
</li>
<li data-start="915" data-end="971">
<p data-start="917" data-end="971">Add educational examples and performance comparisons</p>
</li>
<li data-start="972" data-end="1062">
<p data-start="974" data-end="1062">Follow repository guidelines: type hints, error handling, and pure Python implementation</p>
</li>
</ul>
<hr data-start="1064" data-end="1067">
<h3 data-start="1069" data-end="1090">Technical Details</h3>
<p data-start="1092" data-end="1119"><strong data-start="1092" data-end="1119">Algorithms Implemented:</strong></p>
<div class="_tableContainer_1rjym_1"><div tabindex="-1" class="group _tableWrapper_1rjym_13 flex w-fit flex-col-reverse">
Algorithm | Key Concept
-- | --
SGD | θ = θ - α∇θ (basic gradient descent)
MomentumSGD | v = βv + (1-β)∇θ, θ = θ - αv
NAG | Uses lookahead gradients for better convergence
Adagrad | Adapts learning rates per parameter
Adam | Combines momentum and adaptive learning rates

</div></div>
<p data-start="1739" data-end="1763"><strong data-start="1739" data-end="1763">Directory Structure:</strong></p>
<pre class="overflow-visible!" data-start="1765" data-end="2387"><div class="contain-inline-size rounded-2xl relative bg-token-sidebar-surface-primary"><div class="sticky top-9"><div class="absolute end-0 bottom-0 flex h-9 items-center pe-2"><div class="bg-token-bg-elevated-secondary text-token-text-secondary flex items-center gap-4 rounded-sm px-2 font-sans text-xs"></div></div></div><div class="overflow-y-auto p-4" dir="ltr"><code class="whitespace-pre!"><span><span>neural_network/optimizers/
├── __init__.py                  </span><span><span class="hljs-comment"># Package initialization</span></span><span>
├── README.md                    </span><span><span class="hljs-comment"># Comprehensive documentation</span></span><span>
├── base_optimizer.py            </span><span><span class="hljs-comment"># Abstract base class</span></span><span>
├── sgd.py                       </span><span><span class="hljs-comment"># Stochastic Gradient Descent</span></span><span>
├── momentum_sgd.py              </span><span><span class="hljs-comment"># SGD with Momentum</span></span><span>
├── nag.py                       </span><span><span class="hljs-comment"># Nesterov Accelerated Gradient</span></span><span>
├── adagrad.py                   </span><span><span class="hljs-comment"># Adagrad optimizer</span></span><span>
├── adam.py                      </span><span><span class="hljs-comment"># Adam optimizer</span></span><span>
├── test_optimizers.py           </span><span><span class="hljs-comment"># Comprehensive test suite</span></span><span>
└── IMPLEMENTATION_SUMMARY.md    </span><span><span class="hljs-comment"># Technical implementation details</span></span><span>
</span></span></code></div></div></pre>
<hr data-start="2389" data-end="2392">
<h3 data-start="2394" data-end="2414">Testing Coverage</h3>
<ul data-start="2416" data-end="2559">
<li data-start="2416" data-end="2451">
<p data-start="2418" data-end="2451">61 doctests with 100% pass rate</p>
</li>
<li data-start="2452" data-end="2485">
<p data-start="2454" data-end="2485">Error handling for edge cases</p>
</li>
<li data-start="2486" data-end="2525">
<p data-start="2488" data-end="2525">Multi-dimensional parameter support</p>
</li>
<li data-start="2526" data-end="2559">
<p data-start="2528" data-end="2559">Performance comparison examples</p>
</li>
</ul>
<hr data-start="2561" data-end="2564">
<h3 data-start="2566" data-end="2590">Describe Your Change</h3>
<ul class="contains-task-list" data-start="2592" data-end="2775">
<li class="task-list-item" data-start="2592" data-end="2616">
<p data-start="2598" data-end="2616"><input disabled="" type="checkbox" checked=""> Add an algorithm</p>
</li>
<li class="task-list-item" data-start="2617" data-end="2667">
<p data-start="2623" data-end="2667"><input disabled="" type="checkbox"> Fix a bug or typo in an existing algorithm</p>
</li>
<li class="task-list-item" data-start="2668" data-end="2748">
<p data-start="2674" data-end="2748"><input disabled="" type="checkbox" checked=""> Add or change doctests (note: no mixed code/test changes in a single PR)</p>
</li>
<li class="task-list-item" data-start="2749" data-end="2775">
<p data-start="2755" data-end="2775"><input disabled="" type="checkbox" checked=""> Documentation change</p>
</li>
</ul>
<hr data-start="2777" data-end="2780">
<h3 data-start="2782" data-end="2795">Checklist</h3>
<ul class="contains-task-list" data-start="2797" data-end="3930">
<li class="task-list-item" data-start="2797" data-end="2904">
<p data-start="2803" data-end="2904"><input disabled="" type="checkbox" checked=""> I have read <a data-start="2815" data-end="2901" rel="noopener" target="_new" class="decorated-link" href="https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md">CONTRIBUTING.md<span aria-hidden="true" class="ms-0.5 inline-block align-middle leading-none"><svg width="20" height="20" viewBox="0 0 20 20" fill="currentColor" xmlns="http://www.w3.org/2000/svg" data-rtl-flip="" class="block h-[0.75em] w-[0.75em] stroke-current stroke-[0.75]"><path d="M14.3349 13.3301V6.60645L5.47065 15.4707C5.21095 15.7304 4.78895 15.7304 4.52925 15.4707C4.26955 15.211 4.26955 14.789 4.52925 14.5293L13.3935 5.66504H6.66011C6.29284 5.66504 5.99507 5.36727 5.99507 5C5.99507 4.63273 6.29284 4.33496 6.66011 4.33496H14.9999L15.1337 4.34863C15.4369 4.41057 15.665 4.67857 15.665 5V13.3301C15.6649 13.6973 15.3672 13.9951 14.9999 13.9951C14.6327 13.9951 14.335 13.6973 14.3349 13.3301Z"></path></svg></span></a>.</p>
</li>
<li class="task-list-item" data-start="2905" data-end="2975">
<p data-start="2911" data-end="2975"><input disabled="" type="checkbox" checked=""> This pull request is all my own work — I have not plagiarized.</p>
</li>
<li class="task-list-item" data-start="2976" data-end="3062">
<p data-start="2982" data-end="3062"><input disabled="" type="checkbox" checked=""> I know that pull requests will not be merged if they fail the automated tests.</p>
</li>
<li class="task-list-item" data-start="3063" data-end="3168">
<p data-start="3069" data-end="3168"><input disabled="" type="checkbox"> This PR only changes one algorithm file. (For multiple algorithms, separate PRs are recommended.)</p>
</li>
<li class="task-list-item" data-start="3169" data-end="3238">
<p data-start="3175" data-end="3238"><input disabled="" type="checkbox" checked=""> All new Python files are placed inside an existing directory.</p>
</li>
<li class="task-list-item" data-start="3239" data-end="3300">
<p data-start="3245" data-end="3300"><input disabled="" type="checkbox" checked=""> All filenames are lowercase with no spaces or dashes.</p>
</li>
<li class="task-list-item" data-start="3301" data-end="3370">
<p data-start="3307" data-end="3370"><input disabled="" type="checkbox" checked=""> All functions and variables follow Python naming conventions.</p>
</li>
<li class="task-list-item" data-start="3371" data-end="3468">
<p data-start="3377" data-end="3468"><input disabled="" type="checkbox" checked=""> All functions include Python <a data-start="3406" data-end="3465" class="decorated-link" rel="noopener" target="_new" href="https://docs.python.org/3/library/typing.html">type hints<span aria-hidden="true" class="ms-0.5 inline-block align-middle leading-none"><svg width="20" height="20" viewBox="0 0 20 20" fill="currentColor" xmlns="http://www.w3.org/2000/svg" data-rtl-flip="" class="block h-[0.75em] w-[0.75em] stroke-current stroke-[0.75]"><path d="M14.3349 13.3301V6.60645L5.47065 15.4707C5.21095 15.7304 4.78895 15.7304 4.52925 15.4707C4.26955 15.211 4.26955 14.789 4.52925 14.5293L13.3935 5.66504H6.66011C6.29284 5.66504 5.99507 5.36727 5.99507 5C5.99507 4.63273 6.29284 4.33496 6.66011 4.33496H14.9999L15.1337 4.34863C15.4369 4.41057 15.665 4.67857 15.665 5V13.3301C15.6649 13.6973 15.3672 13.9951 14.9999 13.9951C14.6327 13.9951 14.335 13.6973 14.3349 13.3301Z"></path></svg></span></a>.</p>
</li>
<li class="task-list-item" data-start="3469" data-end="3583">
<p data-start="3475" data-end="3583"><input disabled="" type="checkbox" checked=""> All functions have <a data-start="3494" data-end="3552" rel="noopener" target="_new" class="decorated-link" href="https://docs.python.org/3/library/doctest.html">doctests<span aria-hidden="true" class="ms-0.5 inline-block align-middle leading-none"><svg width="20" height="20" viewBox="0 0 20 20" fill="currentColor" xmlns="http://www.w3.org/2000/svg" data-rtl-flip="" class="block h-[0.75em] w-[0.75em] stroke-current stroke-[0.75]"><path d="M14.3349 13.3301V6.60645L5.47065 15.4707C5.21095 15.7304 4.78895 15.7304 4.52925 15.4707C4.26955 15.211 4.26955 14.789 4.52925 14.5293L13.3935 5.66504H6.66011C6.29284 5.66504 5.99507 5.36727 5.99507 5C5.99507 4.63273 6.29284 4.33496 6.66011 4.33496H14.9999L15.1337 4.34863C15.4369 4.41057 15.665 4.67857 15.665 5V13.3301C15.6649 13.6973 15.3672 13.9951 14.9999 13.9951C14.6327 13.9951 14.335 13.6973 14.3349 13.3301Z"></path></svg></span></a> that pass automated testing.</p>
</li>
<li class="task-list-item" data-start="3584" data-end="3686">
<p data-start="3590" data-end="3686"><input disabled="" type="checkbox" checked=""> All new algorithms include at least one URL to an authoritative explanation (e.g., Wikipedia).</p>
</li>
<li class="task-list-item" data-start="3687" data-end="3930">
<p data-start="3693" data-end="3930"><input disabled="" type="checkbox" checked=""> If this pull request resolves one or more issues, the description includes the issue number(s) with a <a data-start="3795" data-end="3913" class="decorated-link" rel="noopener" target="_new" href="https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue">closing keyword<span aria-hidden="true" class="ms-0.5 inline-block align-middle leading-none"><svg width="20" height="20" viewBox="0 0 20 20" fill="currentColor" xmlns="http://www.w3.org/2000/svg" data-rtl-flip="" class="block h-[0.75em] w-[0.75em] stroke-current stroke-[0.75]"><path d="M14.3349 13.3301V6.60645L5.47065 15.4707C5.21095 15.7304 4.78895 15.7304 4.52925 15.4707C4.26955 15.211 4.26955 14.789 4.52925 14.5293L13.3935 5.66504H6.66011C6.29284 5.66504 5.99507 5.36727 5.99507 5C5.99507 4.63273 6.29284 4.33496 6.66011 4.33496H14.9999L15.1337 4.34863C15.4369 4.41057 15.665 4.67857 15.665 5V13.3301C15.6649 13.6973 15.3672 13.9951 14.9999 13.9951C14.6327 13.9951 14.335 13.6973 14.3349 13.3301Z"></path></svg></span></a>: <code data-start="3915" data-end="3929">Fixes #13662</code>.</p></li></ul>
</body>
</html>Neural Network Optimizers Module

This PR introduces a comprehensive neural network optimizers module that implements five widely used optimization algorithms for machine learning and deep learning.
The primary goal is to enhance the educational value of the repository by including well-documented, tested, and modular implementations.

Fixes #13662

What's Added

Add SGD (Stochastic Gradient Descent) optimizer

Add MomentumSGD with momentum acceleration

Add NAG (Nesterov Accelerated Gradient) optimizer

Add Adagrad with adaptive learning rates

Add Adam optimizer combining momentum and RMSprop

Include 61 comprehensive doctests (all passing)

Add abstract BaseOptimizer for a consistent interface

Include detailed mathematical documentation

Add educational examples and performance comparisons

Follow repository guidelines: type hints, error handling, and pure Python implementation

Technical Details

Algorithms Implemented:

Algorithm	Key Concept
SGD	θ = θ - α∇θ (basic gradient descent)
MomentumSGD	v = βv + (1-β)∇θ, θ = θ - αv
NAG	Uses lookahead gradients for better convergence
Adagrad	Adapts learning rates per parameter
Adam	Combines momentum and adaptive learning rates

Directory Structure:

neural_network/optimizers/
├── __init__.py                  # Package initialization
├── README.md                    # Comprehensive documentation
├── base_optimizer.py            # Abstract base class
├── sgd.py                       # Stochastic Gradient Descent
├── momentum_sgd.py              # SGD with Momentum
├── nag.py                       # Nesterov Accelerated Gradient
├── adagrad.py                   # Adagrad optimizer
├── adam.py                      # Adam optimizer
├── test_optimizers.py           # Comprehensive test suite
└── IMPLEMENTATION_SUMMARY.md    # Technical implementation details

Testing Coverage

61 doctests with 100% pass rate

Error handling for edge cases

Multi-dimensional parameter support

Performance comparison examples

Describe Your Change

 Add an algorithm

 Fix a bug or typo in an existing algorithm

 Add or change doctests (note: no mixed code/test changes in a single PR)

 Documentation change

Checklist

 I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md)
.

 This pull request is all my own work — I have not plagiarized.

 I know that pull requests will not be merged if they fail the automated tests.

 This PR only changes one algorithm file. (For multiple algorithms, separate PRs are recommended.)

 All new Python files are placed inside an existing directory.

 All filenames are lowercase with no spaces or dashes.

 All functions and variables follow Python naming conventions.

 All functions include Python [type hints](https://docs.python.org/3/library/typing.html)
.

 All functions have [doctests](https://docs.python.org/3/library/doctest.html)
 that pass automated testing.

 All new algorithms include at least one URL to an authoritative explanation (e.g., Wikipedia).

 If this pull request resolves one or more issues, the description includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
: Fixes #13662.

### CODE REVIEW

**Comment by shretadas** [comment]
Thank you for the review and the guidance.

I have updated the PR description and checklist to reflect the items that are applicable to this submission. I understand only “[x]” is accepted for marking a point as completed. 

I have two options and I’d like to know which you prefer:
1) If possible, please re-open this PR so I can push a follow-up commit that only updates the checklist and minor docs; or
2) If you prefer a fresh PR, I will open a new PR with the corrected description and the same implementation files.

Below is the updated PR description (already includes Fixes #13662). I will update the PR body now; please let me know whether to re-open this one or if I should create a new PR. Thank you for the help.
